### ISSUE TITLE
Implement data-driven styling for line-dash-array

### ISSUE THREAD

**Issue by a-maggioni** [maplibre/maplibre-gl-js#1235]
I would like to assign different values to "line-dash-array" property based on the actual data.
Example:
`
'line-dasharray': [
      "case", 
      ['==', ['get', "my_prop"], 'my_value'],
      ['literal', [0,0]],
      ['literal', [2,3]]
]
`

This feature has already been added to Mapbox GL JS >= 2.3.0.
I know that there is a workaround (i.e. duplicate the layer and use a filter) but it would be nice to have this feature also in maplibre.

**Comment by HarelM** [comment]
Fell free to push this forward and submit a PR. :-)
Note that it's forbidden to copy code from Mapbox v2 and above.

**Comment by teifler** [comment]
> I would like to assign different values to "line-dash-array" property based on the actual data. Example: `'line-dasharray': [ "case", ['==', ['get', "my_prop"], 'my_value'], ['literal', [0,0]], ['literal', [2,3]] ]`
> 
> This feature has already been added to Mapbox GL JS >= 2.3.0. I know that there is a workaround (i.e. duplicate the layer and use a filter) but it would be nice to have this feature also in maplibre.

If possible could you post a sample of the workaround?

Thanks! :))

**Comment by pnorman** [comment]
> Also will this PR get merged? We would really appropriate that feature

There is no PR - this is an issue.

**Comment by a-maggioni** [comment]
> > I would like to assign different values to "line-dash-array" property based on the actual data. Example: `'line-dasharray': [ "case", ['==', ['get', "my_prop"], 'my_value'], ['literal', [0,0]], ['literal', [2,3]] ]`
> > This feature has already been added to Mapbox GL JS >= 2.3.0. I know that there is a workaround (i.e. duplicate the layer and use a filter) but it would be nice to have this feature also in maplibre.
> 
> If possible could you post a sample of the workaround?
> 
> Also will this PR get merged? We would really appropriate that feature
> 
> Thanks! :))

You can create two or more layers with different "line-dasharray" values and set a filter to display one layer or the other based on the actual data.
Example:
`"layers": [
    {
      "id": "layer-1",
      "type": "line",
      "metadata": {},
      "source": "sourceId",
      "source-layer": "sourceLayerId",
      "minzoom": 1,
      **"filter": ["all", ["==", "my_prop", "my_value"]],**
      "layout": {"visibility": "visible"},
      "paint": {
        "line-color": "rgba(0, 0, 255, 1)",
        "line-width": 1,
        **"line-dasharray": ["literal", [1, 1]]**
      }
    },
    {
      "id": "layer-2",
      "type": "line",
      "metadata": {},
      "source": "sourceId",
      "source-layer": "sourceLayerId",
      "minzoom": 1,
      **"filter": ["all", ["!=", "my_prop", "my_value"]],**
      "layout": {"visibility": "visible"},
      "paint": {
        "line-color": "rgba(0, 0, 255, 1)",
        "line-width": 1,
        **"line-dasharray": ["literal", [1, 3]]**
      }
    }
  ]`

**Comment by ZeLonewolf** [comment]
I would like to attempt a PR for this, as I believe it would enable a significant performance improvement in Americana.  If anyone has done any work on this or is currently working on this, please reach out.

**Comment by tarwin** [comment]
Would really love this as well. Having the same issue as others. Of course could do the workaround but adds overhead. Thank you all for your amazing work !

**Comment by teifler** [comment]
> You can create two or more layers with different "line-dasharray" values and set a filter to display one layer or the other based on the actual data. Example:
> 
> ```
> "layers": [
>     {
>       "id": "layer-1",
>       "type": "line",
>       "metadata": {},
>       "source": "sourceId",
>       "source-layer": "sourceLayerId",
>       "minzoom": 1,
>       "filter": ["all", ["==", "my_prop", "my_value"]],
>       "layout": {"visibility": "visible"},
>       "paint": {
>         "line-color": "rgba(0, 0, 255, 1)",
>         "line-width": 1,
>         "line-dasharray": ["literal", [1, 1]]
>       }
>     },
>     {
>       "id": "layer-2",
>       "type": "line",
>       "metadata": {},
>       "source": "sourceId",
>       "source-layer": "sourceLayerId",
>       "minzoom": 1,
>       "filter": ["all", ["!=", "my_prop", "my_value"]],
>       "layout": {"visibility": "visible"},
>       "paint": {
>         "line-color": "rgba(0, 0, 255, 1)",
>         "line-width": 1,
>         "line-dasharray": ["literal", [1, 3]]
>       }
>     }
> ]
> ```
This works like a charm thanks!

**Comment by berkayoruc** [comment]
line-cap also need data driven styling

**Comment by HarelM** [comment]
Assigning a XL bounty as requested by @Samarth1696 and approved by the MapLibre board.

**Comment by Samarth1696** [comment]
@HarelM I would like to work on this.

**Comment by JacobArrow** [comment]
Will this also allow gradients for line-dash-array?

**Comment by Samarth1696** [comment]
I don't think that would be the scope of this issue. I have multiple ideas regarding lines like performance and other things. As of now, this implementation will not inherently allow gradients specifically for `line-dash`. If you need to apply gradients to dashed lines, additional work or a separate feature implementation might be necessary to support that functionality.

If gradient support for dashed lines is important for your project, I recommend opening a new issue or feature request after the implementation of data-driven for `line-dash` to discuss it further with the community.

**Comment by Michal-Sw-airspace-intelligence** [comment]
> Will this also allow gradients for line-dash-array?

Created a feature request for this https://github.com/maplibre/maplibre-gl-js/issues/5082

**Comment by bernatfortet** [comment]
@Samarth1696 did you have a chance to implement this?

**Comment by Samarth1696** [comment]
Sorry, got busy with other things but will try to complete it in this year only.

**Comment by neodescis** [comment]
It would be great to see this done, especially as Mapbox Draw 1.5.0 now provides default styling that is no longer compatible with MapLibre because of this ([link](https://github.com/mapbox/mapbox-gl-draw/issues/1357#issuecomment-2705533280)).

**Comment by Samarth1696** [comment]
I have no issue.
@HarelM Pls unassign me.

**Comment by HarelM** [comment]
@lucaswoj I have assigned this issue to you. 
Please let me know if you decide not to pursue it.
Also let me know up front if the bounty's size is OK.
If there are other issue which you would like to work on and do not have a bounty do let me know. 
Thanks!!

**Comment by jo-chemla** [comment]
Hi there, probably not the right place to ask such a question, but since I saw this mentioned when parsing this thread I wanted to be sure. 

>  Note that it's forbidden to copy code from Mapbox v2 and above.

I'm submitting a (still draft) PR https://github.com/mapbox/mapbox-gl-js/pull/13535 to mapbox to avoid flickers when switching raster-tile-source layer TMS url via `setTiles` - useful when using historical tiled rasters, to avoid tiles popping in and having a smooth cross-fade between previous and new tiles, per-tile. 

Reason for this comment is I just wanted to make sure that I will not be bound to mapbox licensing terms, and have the right since I'm the author of that patch to also port it to maplibre if deemed useful. Thanks!

**Comment by HarelM** [comment]
@jo-chemla I think it would be safer/easier if this would land here first and then at mapbox.
The other way around might need the intervention of a legal department, which I'd like to avoid as much as possible. 

If not, we can try and see what can be done to solve this. 
But I believe it's not related to this thread, so I would advise to open a new discussion and talk about it there. 
It would also serve as a paper trail, just in case...
